### ISSUE TITLE
L2 regularization for constant optimization

### ISSUE THREAD

**Issue by Smantii** [heal-research/pyoperon#23]
Hi,

is there a way to penalize the magnitude of the constants (via, e.g., L2 regularization)? I am trying to fit a `SymbolicRegressor` with some noisy data and sometimes I get very large values for some constants.
I looked inside the library and it seems that it is possible to choose, as constant optimizer, `adamax` and `amsgrad`. In the `pytorch` implementation of these optimizers there is a parameter `weight_decay` for the $L^2$ regularization and I was guessing if it is possible to do so also for `pyoperon`.

**Comment by foolnotion** [comment]
Hi,

Regularization is currently not supported but we can look into adding it. Is there an official paper or link to the pytorch implementation?

**Comment by Smantii** [comment]
Hi,

here are some links to the pytorch docs:
- [adamax](https://pytorch.org/docs/stable/generated/torch.optim.Adamax.html)
- [Adam](https://pytorch.org/docs/stable/generated/torch.optim.Adam.html)

**Comment by gkronber** [comment]
Note, that regularization can be misleading or unwanted for expressions with nonlinear parameters that GP might produce $f(x) / \theta$ or $f(x) \exp(-\theta)$

Regularization of the raw parameters may lead to an evolutionary advantage of nonlinear transformations of parameters to allow 'virtual' large parameters.

**Comment by Smantii** [comment]
Can you explain a bit more this point? I am not sure that I have completely understood, sorry

**Comment by Smantii** [comment]
You are right, but this is true only for specific primitives right? In other words, if the primitive set is, e.g., 

```python
allowed_symbols = "add,sub,mul,sin,cos,constant,variable"
```
there is no such problem, right?

**Comment by gkronber** [comment]
Yes, in this case this is less of a problem.

**Comment by foolnotion** [comment]
> Hi,
> 
> here are some links to the pytorch docs:
> 
>     * [adamax](https://pytorch.org/docs/stable/generated/torch.optim.Adamax.html)
> 
>     * [Adam](https://pytorch.org/docs/stable/generated/torch.optim.Adam.html)

By the way, here are the implementations available in Operon:
https://github.com/heal-research/operon/blob/main/include/operon/optimizer/solvers/sgd.hpp

Currently not all are exposed in the Python wrapper, but it's trivial to add them.
